android/src/main/java/com/oblador/keychain/KeychainModule.java

```diff
@@ -13,291 +20,586 @@
+import com.oblador.keychain.cipherStorage.CipherStorageKeystoreAesCbc;
+import com.oblador.keychain.cipherStorage.CipherStorageKeystoreRsaEcb;
+import com.oblador.keychain.cipherStorage.CipherStorageKeystoreRsaEcb.NonInteractiveHandler;
 import com.oblador.keychain.exceptions.CryptoFailedException;
 import com.oblador.keychain.exceptions.EmptyParameterException;
 import com.oblador.keychain.exceptions.KeyStoreAccessException;
 
 import java.util.HashMap;
 import java.util.Map;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
 
-import javax.annotation.Nullable;
+import static com.oblador.keychain.SecurityLevel.ANY;
+import static com.oblador.keychain.SecurityLevel.SECURE_HARDWARE;
+import static com.oblador.keychain.SecurityLevel.SECURE_SOFTWARE;
 
+@SuppressWarnings({"unused", "WeakerAccess"})
 public class KeychainModule extends ReactContextBaseJavaModule {
+  //region Constants
+  public static final String KEYCHAIN_MODULE = "RNKeychainManager";
+  public static final String FINGERPRINT_SUPPORTED_NAME = "Fingerprint";
+  public static final String EMPTY_STRING = "";
+  public static final String ACCESS_CONTROL_BIOMETRY_ANY = "BiometryAny";
+  public static final String ACCESS_CONTROL_BIOMETRY_CURRENT_SET = "BiometryCurrentSet";
+
+  @interface Maps {
+    String SERVICE = "service";
+    String USERNAME = "username";
+    String PASSWORD = "password";
+  }
+
+  @interface Errors {
+    String E_EMPTY_PARAMETERS = "E_EMPTY_PARAMETERS";
+    String E_CRYPTO_FAILED = "E_CRYPTO_FAILED";
+    String E_KEYSTORE_ACCESS_ERROR = "E_KEYSTORE_ACCESS_ERROR";
+    String E_SUPPORTED_BIOMETRY_ERROR = "E_SUPPORTED_BIOMETRY_ERROR";
+    /** Raised for unexpected errors. */
+    String E_UNKNOWN_ERROR = "E_UNKNOWN_ERROR";
+  }
+
+  //endregion
+
+  //region Members
+  /** Name-to-instance lookup  map. */
+  private final Map<String, CipherStorage> cipherStorageMap = new HashMap<>();
+  /** Shared preferences storage. */
+  private final PrefsStorage prefsStorage;
+  //endregion
+
+  public KeychainModule(@NonNull final ReactApplicationContext reactContext) {
+    super(reactContext);
+    prefsStorage = new PrefsStorage(reactContext);
+
+    addCipherStorageToMap(new CipherStorageFacebookConceal(reactContext));
+    addCipherStorageToMap(new CipherStorageKeystoreAesCbc());
+
+    // we have a references to newer api that will fail load of app classes in old androids OS
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+      addCipherStorageToMap(new CipherStorageKeystoreRsaEcb());
+    }
+  }
+
+  //region Overrides
+
+  /** {@inheritDoc} */
+  @Override
+  @NonNull
+  public String getName() {
+    return KEYCHAIN_MODULE;
+  }
+
+  /** {@inheritDoc} */
+  @NonNull
+  @Override
+  public Map<String, Object> getConstants() {
+    final Map<String, Object> constants = new HashMap<>();
+
+    constants.put(ANY.jsName(), ANY.name());
+    constants.put(SECURE_SOFTWARE.jsName(), SECURE_SOFTWARE.name());
```

Comment:
To follow code style used in this library:
```suggestion
    constants.put(SecurityLevel.SECURE_SOFTWARE.jsName(), SecurityLevel.SECURE_SOFTWARE.name());
```